scrape/target.go

```diff
@@ -507,3 +459,85 @@ func targetsFromGroup(tg *targetgroup.Group, cfg *config.ScrapeConfig) ([]*Targe
 	}
 	return targets, failures
 }
+
+// Check after relabel
+func checkAfterRelabel(lset labels.Labels, cfg *config.ScrapeConfig) (err error) {
+	// Check address
+	addr := lset.Get(model.AddressLabel)
+	if addr == "" {
+		return errors.New("no address")
+	}
+
+	if err := config.CheckTargetAddress(model.LabelValue(addr)); err != nil {
+		return err
+	}
+
+	var interval string
+	var intervalDuration model.Duration
+	if interval = lset.Get(model.ScrapeIntervalLabel); interval != cfg.ScrapeInterval.String() {
+		intervalDuration, err = model.ParseDuration(interval)
+		if err != nil {
+			return errors.Errorf("error parsing scrape interval: %v", err)
+		}
+		if time.Duration(intervalDuration) == 0 {
+			return errors.New("scrape interval cannot be 0")
+		}
+	}
+
+	var timeout string
+	var timeoutDuration model.Duration
+	if timeout = lset.Get(model.ScrapeTimeoutLabel); timeout != cfg.ScrapeTimeout.String() {
+		timeoutDuration, err = model.ParseDuration(timeout)
+		if err != nil {
+			return errors.Errorf("error parsing scrape timeout: %v", err)
+		}
+		if time.Duration(timeoutDuration) == 0 {
+			return errors.New("scrape timeout cannot be 0")
+		}
+	}
+
+	if timeoutDuration > intervalDuration {
+		return errors.Errorf("scrape timeout cannot be greater than scrape interval (%q > %q)", timeout, interval)
+	}
+
+	return nil
+}
+
+// populateLabels reuse doRelabel to build a label set from the given label set and scrape configuration and check after relabel.
+func populateLabels(lset labels.Labels, cfg *config.ScrapeConfig) (res, orig labels.Labels, err error) {
```

Comment:
Why are we changing this function instead of exposing it?

Reply:
The latest commit is just changing for reuse AddPortIfNecessary func.